### PR TITLE
Exchange issuance zero ex extended integration test

### PR DESCRIPTION
**Pull request by ckoopmann**

### CODE REVIEW

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+type SetTokenScenario = {
+  setToken: Address;
+  controller: Address;
+  issuanceModule: Address;
+};
+
+type TokenName = "SimpleToken" | "DPI";
+
+function logIfVerboseMode(...args: any[]) {
```
Comment (FlattestWhite) [inline test/integration/exchangeIssuanceZeroExIntegration.spec.ts:31]: `logVerbose`?

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+    let deployer: DeployHelper;
+
+    // Contract Addresses
+    let wethAddress: Address;
+    let wbtcAddress: Address;
+    let daiAddress: Address;
+    let zeroExProxyAddress: Address;
+    let controllerAddress: Address;
+    let issuanceModuleAddress: Address;
+    const DPI_ADDRESS: Address = "0x1494ca1f11d487c2bbe4543e90080aeba4ba3c2b";
```
Comment (FlattestWhite) [inline test/integration/exchangeIssuanceZeroExIntegration.spec.ts:49]: Move const upwards above the `lets` or instantiate like the other addresses below?

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+                setToken: SetToken,
+                inputTokenAddress: Address,
+                setAmount: number,
+                slippagePercents: number,
+                excludedSources: string | undefined = undefined,
+              ): Promise<[ZeroExSwapQuote, ZeroExSwapQuote[], BigNumber]> {
+                const positions = await setToken.getPositions();
+                const positionQuotes: ZeroExSwapQuote[] = [];
+                let buyAmountWeth = BigNumber.from(0);
+                const slippagePercentage = slippagePercents / 100;
```
Comment (FlattestWhite) [inline test/integration/exchangeIssuanceZeroExIntegration.spec.ts:195]: This is a little confusing to the reader if they didn't understand the 0x api. Perhaps add a comment here saying that 0.5% is represented as 0.005?

**Review by FlattestWhite — COMMENTED** [review]
Look pretty good to me, just had a few nits and comments. Btw is the yarn.lock intended?

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+                const outputQuote = {
+                  sellToken: wethAddress,
+                  buyToken: outputTokenAddress,
+                  swapCallData: outputTokenApiResponse.data,
+                };
+                return [outputQuote, positionQuotes, outputTokenAmount];
+              }
+
+              const initializeSubjectVariables = async (_amountSetToken: number) => {
+                const SLIPPAGE_PERCENTAGE = 50;
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply test/integration/exchangeIssuanceZeroExIntegration.spec.ts:385]: We get "underbought" reverts from the 0x Exchange Proxy

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+              }
+              beforeEach(async () => {
+                await deployExchangeIssuanceZeroEx();
+                await exchangeIssuanceZeroEx.approveSetToken(setToken.address);
+                await initializeSubjectVariables(setTokenAmount);
+                await setToken
+                  .connect(owner.wallet)
+                  .approve(exchangeIssuanceZeroEx.address, subjectAmountSetTokenWei);
+              });
+              it("should consume correct amount of set tokens", async () => {
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply test/integration/exchangeIssuanceZeroExIntegration.spec.ts:421]: True, can add that assertion

**Comment by ckoopmann** [comment]
> Look pretty good to me, just had a few nits and comments. Btw is the yarn.lock intended?

I had to upgrade hardhat since the old version failed with something like "invalid v signature" when running tests agains an up to date fork

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts (hunk 1 above)
Comment (ckoopmann) [reply test/integration/exchangeIssuanceZeroExIntegration.spec.ts:31]: sure is shorter. Adjusted the name

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+                ] = await getRedemptionQuotes(
+                  setToken,
+                  subjectOutputToken.address,
+                  subjectAmountSetToken,
+                  SLIPPAGE_PERCENTAGE,
+                );
+              };
+              async function subject(): Promise<ContractTransaction> {
+                return await exchangeIssuanceZeroEx
+                  .connect(owner.wallet)
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply test/integration/exchangeIssuanceZeroExIntegration.spec.ts:404]: Added a second account "user" that signs everything but the deployment.

File: test/integration/exchangeIssuanceZeroExIntegration.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ckoopmann) [reply test/integration/exchangeIssuanceZeroExIntegration.spec.ts:432]: Added an assertion checking that we got at least the specified minimum amount of output token.

**Pull request merged**